fileshare/transfer.go

```diff
@@ -159,197 +161,338 @@ func GetTransferFileStatus(file *pb.File, in bool) (status string) {
+				// adjust file statuses
+				for _, file := range allFiles {
 					if !isFileCompleted(file) {
 						file.Status = pb.Status_CANCELED
 					}
 				}
 			}
-		case "failed":
-			out.Status = pb.Status(lastState.StatusCode)
+		case norddrop.TransferStateKindFailed:
+			lastState := lastState.Kind.(norddrop.TransferStateKindFailed)
+			status = pb.Status(lastState.Status)
 		}
 	} else {
-		out.Status = getTransferStatus(out.Files)
+		status = getTransferStatusByFiles(allFiles)
 	}
+	return status
+}
 
-	return out
+func getTransferStatusByFiles(files []*pb.File) pb.Status {
+	allCanceled := true
+	allFinished := true
+	hasNoErrors := true
+	hasStarted := false
+	for _, file := range files {
+		if allCanceled && file.Status != pb.Status_CANCELED {
+			allCanceled = false
+		}
+		if allFinished && !isFileCompleted(file) {
+			allFinished = false
+		}
+		if hasNoErrors && checkFileHasErrors(file) {
+			hasNoErrors = false
+		}
+		if !hasStarted && file.Status != pb.Status_REQUESTED {
+			hasStarted = true
+		}
+	}
+
+	switch {
+	case allCanceled:
+		return pb.Status_CANCELED
+	case allFinished && hasNoErrors:
+		return pb.Status_SUCCESS
+	case allFinished && !hasNoErrors:
+		return pb.Status_FINISHED_WITH_ERRORS
+	case hasStarted:
+		return pb.Status_ONGOING
+	default:
+		return pb.Status_REQUESTED
+	}
 }
 
-func libdropFileToInternalFile(in LibdropFile) *pb.File {
-	out := &pb.File{
-		Id:       in.ID,
-		Path:     in.RelativePath,
-		FullPath: filepath.Join(in.BasePath, in.RelativePath),
-		Size:     in.TotalSize,
-		// This only shows the amount from the last status change.
-		// The only up to date source for data transferred are events.
-		// To show the correct value we track ongoing transfers in event manager,
-		// and event manager overwrites values of transferred fields if needed.
-		Transferred: 0,
+func determineTransferPath(allFiles []*pb.File) string {
+	if len(allFiles) == 0 {
+		return ""
 	}
 
-	for _, state := range in.States {
-		// BasePath is provided with the very first "pending" state
-		if state.BasePath != "" {
-			out.FullPath = filepath.Join(state.BasePath, out.FullPath)
-		}
-		// FinalPath is provided with the very last "completed" state, so if it is present, then
-		// it will overwrite the previously constructed path
-		if state.FinalPath != "" {
-			out.FullPath = state.FinalPath
-		}
+	if len(allFiles) == 1 {
+		// tranmitting single file, so the transfer path is full file path
+		return allFiles[0].FullPath
+	}
 
-		if state.BytesReceived != 0 {
-			out.Transferred = state.BytesReceived
-		}
-		if state.BytesSent != 0 {
-			out.Transferred = state.BytesSent
+	filePaths := make([]string, len(allFiles))
+	for i, file := range allFiles {
+		filePaths[i] = file.FullPath
+	}
+
+	transferPath := findLongestCommonPrefix(filePaths)
+	if transferPath == "" || transferPath == "/" {
+		transferPath = "multiple files"
+	}
+
+	return transferPath
+}
+
+func findLongestCommonPrefix(filePaths []string) string {
```

Comment:
Isn't there too many iterations over the list? :smile: 
Also this should prob go in the `string.go` file, maybe we ca reuse it?

didn't tested the bellow impl, but just an idea? :laughing: 
```
func commonPrefix(strs []string) string {
	if len(strs) == 0 {
		return ""
	}

	// Start with the first string as the initial prefix
	prefix := strs[0]

	for _, str := range strs[1:] {
		l := min(len(prefix), len(str))
		for i := 0; i < l; i++ {
			if prefix[i] != str[i] {
				prefix = prefix[:i]
				break
			}
		}
	}

	return prefix
}
```

Reply:
I actually had to get rid of the idea entirely. Files from the transfer are not enough to determine the path because the transfer which is not yet accepted has no concept of a path :cry: